photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala

```diff
@@ -0,0 +1,258 @@
+import com.linkedin.photon.ml.optimization.DistributedOptimizationProblem
+import com.linkedin.photon.ml.optimization.game._
+import com.linkedin.photon.ml.sampler.{BinaryClassificationDownSampler, DefaultDownSampler, DownSampler}
+import com.linkedin.photon.ml.supervised.classification.{LogisticRegressionModel, SmoothedHingeLossLinearSVMModel}
+import com.linkedin.photon.ml.supervised.regression.{LinearRegressionModel, PoissonRegressionModel}
+import com.linkedin.photon.ml.util.Implicits._
+import com.linkedin.photon.ml.util._
+
+/**
+ * Estimator implementation for GAME models.
+ *
+ * @param sc The spark context for the application
+ * @param taskType The training task for the estimator
+ * @param numOuterIterations The number of coordinate descent iterations to perform during estimation (the number of
+ *                           times to pass through the coordinate update sequence)
+ * @param computeVariance Whether to compute the coefficient variances for estimated GAME models
+ * @param logger The logger instance for the application
+ */
+class GameEstimator(
+    val sc: SparkContext,
+    val taskType: TaskType,
+    val numOuterIterations: Int,
+    val computeVariance: Boolean,
+    implicit val logger: Logger) {
+
+  import GameEstimator._
+
+  // 2 types that makes the code more readable
+  type LossFunction = (PointwiseLossFunction) => SingleNodeGLMLossFunction
+  type DistributedLossFunction = (PointwiseLossFunction) => DistributedGLMLossFunction
+
+  private val defaultNormalizationContext: Broadcast[NormalizationContext] = sc.broadcast(NoNormalization())
+
+  /**
+   * Fit a GAME model with the given configuration to the given training data.
+   *
+   * The configuration is used to define one or more 'coordinates' of various types (i.e. fixed effect, random effect,
+   * matrix factorization) which together for the complete mixed effect optimization problem. The coordinates are
+   * updated in a provided order, a provided number of times. For optimum performance, the update order should start
+   * with the most general 'coordinates' and end with the least general - each successive update learning the residuals
+   * of the previous 'coordinates'.
+   *
+   * @param configuration The configuration for the GAME optimization problem
+   * @param updatingSequence The coordinate components of the model to train, and the order in which to train them
+   * @param trainingDataSets The training data sets for each coordinate of the GAME optimization problem
+   * @param trainingEvaluator The evaluator for the training function loss
+   * @param validationDataAndEvaluators Optional validation data and evaluators
+   * @param normalizationContexts Optional normalization contexts
+   * @return A trained GAME model
+   */
+  def fit(
+      configuration: GameModelOptimizationConfiguration,
+      updatingSequence: Seq[CoordinateId],
+      trainingDataSets: Map[CoordinateId, DataSet[_ <: DataSet[_]]],
```

Comment:
This is not quite right. `Estimator#fit` should accept `DataFrame`, not a custom data format. This is the contract in `spark.ml`, and it's also a better design generally speaking. One of the main purposes of organizing things this way in the first place was to generalize on DataFrame/Dataset as input. That's why the conversion was originally happening in this class.